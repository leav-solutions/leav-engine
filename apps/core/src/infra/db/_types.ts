// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
// Copyright LEAV Solutions 2017
// This file is released under LGPL V3

import {AqlQuery} from 'arangojs/lib/cjs/aql-query';
import {IQueryInfos} from '_types/queryInfos';

// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export interface IDbDocument {
    _id: string;
    _key: string;
    _rev: string;
    [key: string]: any;
}

export interface IDbEdge extends IDbDocument {
    _from: string;
    _to: string;
}

export interface IExecute {
    query: string | AqlQuery;
    ctx: IQueryInfos;
    withTotalCount?: boolean;
    attempts?: number;
}

export interface IExecuteWithCount<T = IDbDocument> {
    totalCount: number;
    results: T[];
}

export const isExecuteWithCount = (res: IExecuteWithCount<unknown> | unknown[]): res is IExecuteWithCount<unknown> => {
    return typeof (res as IExecuteWithCount).results !== 'undefined';
};
