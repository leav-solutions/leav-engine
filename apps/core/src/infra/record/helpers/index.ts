// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
export {default as filterTypes} from './filterTypes';
export {default as getClassifyingFiltersVariableQueryPart} from './getClassifyingFiltersVariableQueryPart';
export {default as getSearchVariableName} from './getSearchVariableName';
export {default as getSearchVariablesQueryPart} from './getSearchVariablesQueryPart';
