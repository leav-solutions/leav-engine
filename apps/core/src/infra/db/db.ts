// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IConfig} from '_types/config';

interface IDeps {
    config?: IConfig;
}

const _getDbConnection = (config: IConfig): Database => {
    return new Database({
        url: config.db.url
    });
};

export default function (deps: IDeps): Database {
    const db = _getDbConnection(deps.config);

    return db.database(deps.config.db.name);
}

export const initDb = async (config: IConfig) => {
    const db = _getDbConnection(config);

    const databases = await db.listDatabases();
    const dbExists = databases.reduce((exists, d) => exists || d === config.db.name, false);

    if (!dbExists) {
        await db.createDatabase(config.db.name);
    }

    db.close();
};
