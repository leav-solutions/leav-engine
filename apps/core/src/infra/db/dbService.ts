// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {getCallStack} from '@leav/utils';
import {Database} from 'arangojs';
import {isAqlQuery} from 'arangojs/lib/cjs/aql-query';
import {createHash} from 'crypto';
import {IUtils} from 'utils/utils';
import {IConfig} from '_types/config';
import {IDbProfiler} from '_types/dbProfiler';
import {IDbDocument, IExecute, IExecuteWithCount} from './_types';

const MAX_ATTEMPTS = 10;

export interface IDbService {
    db?: Database;

    /**
     * Execute an AQL query
     * If withTotalCount is set to true, return an object with totalCount and results. Otherwise, just return
     * results straight from DB.
     *
     * @param query
     * @param withTotalCount
     * @param attempts Used when we have to retry a query after a write-write conflict
     * @throws If query fails or we still have a conflict after all attempts
     */
    execute?<T extends IExecuteWithCount<unknown> | unknown[] = IDbDocument[]>(params: IExecute): Promise<T>;

    /**
     * Create a new collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    createCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Delete a collection in database
     *
     * @param name Collection name
     * @param type Document or edge collection?
     * @throws If collection already exists
     */
    dropCollection?(name: string, type?: collectionTypes): Promise<void> | null;

    /**
     * Check if collection already exists
     *
     * @param name
     */
    collectionExists?(name: string): Promise<boolean>;
}

export enum collectionTypes {
    DOCUMENT = 'document',
    EDGE = 'edge'
}

interface IDeps {
    'core.infra.db'?: Database;
    'core.utils'?: IUtils;
    config?: IConfig;
}

export default function ({
    'core.infra.db': db = null,
    'core.utils': utils = null,
    config = null
}: IDeps = {}): IDbService {
    const collectionExists = async function (name: string): Promise<boolean> {
        const collections = await db.listCollections();

        return collections.reduce((exists, c) => exists || c.name === name, false);
    };

    const _sleep = (ms: number): Promise<void> => {
        return new Promise(resolve => setTimeout(resolve, ms));
    };

    return {
        db,
        async execute<T extends IExecuteWithCount<any> | any[] = any[]>({
            query,
            ctx,
            withTotalCount = false,
            attempts = 0
        }: IExecute): Promise<T> {
            try {
                if (config.dbProfiler.enable) {
                    const dbProfiler: IDbProfiler = ctx.dbProfiler ?? {
                        totalCount: 0,
                        uniqueQueriesCount: 0,
                        queries: {}
                    };

                    dbProfiler.totalCount = (dbProfiler.totalCount ?? 0) + 1;

                    // Generate a hash from the query to be able
                    // to group identical queries (exact same query with exact same params)
                    const queryKey = createHash('md5').update(JSON.stringify(query)).digest('base64');

                    if (!dbProfiler.queries) {
                        dbProfiler.queries = {};
                    }

                    const callStack = JSON.stringify(getCallStack(10));

                    dbProfiler.queries[queryKey] = {
                        count: (dbProfiler.queries?.[queryKey]?.count ?? 0) + 1,
                        callers: (dbProfiler.queries?.[queryKey]?.callers ?? new Set()).add(callStack),
                        query
                    };

                    dbProfiler.uniqueQueriesCount = Object.keys(dbProfiler.queries).length;

                    ctx.dbProfiler = dbProfiler;
                }

                // Convert query to AqlQuery if we have a simple query to match query() types
                const queryToRun = isAqlQuery(query)
                    ? {...query}
                    : {
                          query,
                          bindVars: {}
                      };

                const queryOptions = withTotalCount ? {count: true, options: {fullCount: true}} : {};

                const cursor = await db.query(queryToRun, queryOptions);

                const results = await cursor.all();
                return withTotalCount ? {totalCount: cursor.extra.stats.fullCount, results} : results;
            } catch (e) {
                // Handle write-write conflicts: we try the query again with a growing delay between trials.
                // If we reach maximum attempts and still no success, stop it and throw the exception
                // error 1200 === conflict
                if (e.isArangoError && e.errorNum === 1200 && attempts < MAX_ATTEMPTS) {
                    const timeToWait = 2 ** attempts;
                    await _sleep(timeToWait);
                    return this.execute({query, ctx, withTotalCount, attempts: attempts + 1});
                }

                e.message += `\nQuery was: ${JSON.stringify(query).replace(/\\n/g, ' ')}`;
                e.query = query;

                // Response contains circular references which can cause an error when converted to JSON
                // later on. It doesn't contains useful information anyway, so throw it away.
                delete e.response;

                utils.rethrow(e);
            }
        },
        async createCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (await collectionExists(name)) {
                throw new Error(`Collection ${name} already exists`);
            }

            if (type === collectionTypes.EDGE) {
                const collection = db.edgeCollection(name);
                await collection.create();
            } else {
                const collection = db.collection(name);
                await collection.create();
            }
        },
        async dropCollection(name: string, type = collectionTypes.DOCUMENT): Promise<void> {
            if (!(await collectionExists(name))) {
                throw new Error(`Collection ${name} does not exist`);
            }

            const collection = type === collectionTypes.EDGE ? db.edgeCollection(name) : db.collection(name);
            await collection.drop();
        },
        collectionExists
    };
}
