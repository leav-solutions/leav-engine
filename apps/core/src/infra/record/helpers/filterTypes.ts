// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {AttributeCondition, IRecordFilterOption, TreeCondition} from '../../../_types/record';

export interface IFilterTypesHelper {
    isAttributeFilter(filter: IRecordFilterOption): boolean;
    isClassifyingFilter(filter: IRecordFilterOption): boolean;
    isCountFilter(filter: IRecordFilterOption): boolean;
}

export default function(): IFilterTypesHelper {
    const _isAttributeFilter = (filter: IRecordFilterOption) => {
        return (filter as IRecordFilterOption).condition in AttributeCondition;
    };

    const _isClassifyingFilter = (filter: IRecordFilterOption) => {
        return (filter as IRecordFilterOption).condition in TreeCondition;
    };

    const _isCountFilter = (filter: IRecordFilterOption) => {
        return [
            AttributeCondition.VALUES_COUNT_EQUAL,
            AttributeCondition.VALUES_COUNT_GREATER_THAN,
            AttributeCondition.VALUES_COUNT_LOWER_THAN,
            AttributeCondition.IS_EMPTY,
            AttributeCondition.IS_NOT_EMPTY
        ].includes(filter.condition as AttributeCondition);
    };

    return {
        isAttributeFilter: _isAttributeFilter,
        isClassifyingFilter: _isClassifyingFilter,
        isCountFilter: _isCountFilter
    };
}
