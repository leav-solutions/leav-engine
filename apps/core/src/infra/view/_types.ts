// Copyright LEAV Solutions 2017 until 2023/11/05, Copyright Aristid from 2023/11/06
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IDbService} from 'infra/db/dbService';
import {IDbUtils} from 'infra/db/dbUtils';
import {IList} from '_types/list';
import {IGetCoreEntitiesParams} from '_types/shared';
import {IQueryInfos} from '../../_types/queryInfos';

export const VIEWS_COLLECTION_NAME = 'core_views';

export interface IViewRepo {
    createView(view: IView, ctx: IQueryInfos): Promise<IView>;
    updateView(view: IView, ctx: IQueryInfos): Promise<IView>;
    getViews(params: IGetCoreEntitiesParams, ctx: IQueryInfos): Promise<IList<IView>>;
    deleteView(viewId: string, ctx: IQueryInfos): Promise<IView>;
}

export interface IViewRepoDeps {
    'core.infra.db.dbService': IDbService;
    'core.infra.db.dbUtils': IDbUtils;
}
