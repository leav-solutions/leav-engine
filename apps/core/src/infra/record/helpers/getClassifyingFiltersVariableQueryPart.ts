// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {GeneratedAqlQuery, literal} from 'arangojs/aql';
import {IDbService} from 'infra/db/dbService';
import {IRecordFilterOption} from '_types/record';
import {getEdgesCollectionName, getFullNodeId, getRootId} from '../../../infra/tree/helpers/utils';
import {MAX_TREE_DEPTH} from '../../../infra/tree/treeRepo';
import {NODE_LIBRARY_ID_FIELD, NODE_RECORD_ID_FIELD} from '../../../infra/tree/_types';
import {IFilterTypesHelper} from './filterTypes';

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.record.helpers.filterTypes'?: IFilterTypesHelper;
}

export type GetClassifyingFiltersVariableQueryPart = (filter: IRecordFilterOption) => GeneratedAqlQuery;

export default function ({
    'core.infra.record.helpers.filterTypes': filterTypesHelper = null,
    'core.infra.db.dbService': dbService = null
}: IDeps): GetClassifyingFiltersVariableQueryPart {
    return filter => {
        if (!filterTypesHelper.isClassifyingFilter(filter)) {
            return null;
        }

        const collec = dbService.db.collection(getEdgesCollectionName(filter.treeId));

        const startingNode = filter.value
            ? getFullNodeId(String(filter.value), filter.treeId)
            : getRootId(filter.treeId);

        const queryPart = aql`
            FOR v, e IN 1..${MAX_TREE_DEPTH} OUTBOUND ${startingNode}
                ${collec}
                LET record = DOCUMENT(
                    v.${literal(NODE_LIBRARY_ID_FIELD)},
                    v.${literal(NODE_RECORD_ID_FIELD)}
                )
                RETURN record._id
        `;

        return queryPart;
    };
}
