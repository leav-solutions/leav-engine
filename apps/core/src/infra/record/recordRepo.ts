// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {IQueryInfos} from '_types/queryInfos';
import {GeneratedAqlQuery, aql, join, literal} from 'arangojs/aql';
import {GetConditionPart} from 'infra/attributeTypes/helpers/getConditionPart';
import {IDbDocument, IExecuteWithCount} from 'infra/db/_types';
import {GetSearchQuery} from 'infra/indexation/helpers/getSearchQuery';
import {
    CursorDirection,
    ICursorPaginationParams,
    IListWithCursor,
    IPaginationCursors,
    IPaginationParams
} from '../../_types/list';
import {
    AttributeCondition,
    IRecord,
    IRecordFilterOption,
    IRecordSort,
    Operator,
    TreeCondition
} from '../../_types/record';
import {IAttributeRepo} from '../attribute/attributeRepo';
import {IAttributeTypesRepo} from '../attributeTypes/attributeTypesRepo';
import {IDbService} from '../db/dbService';
import {IDbUtils} from '../db/dbUtils';
import {IFilterTypesHelper} from './helpers/filterTypes';
import {GetSearchVariableName} from './helpers/getSearchVariableName';
import {GetSearchVariablesQueryPart} from './helpers/getSearchVariablesQueryPart';

export interface IFindRequestResult {
    initialVars: GeneratedAqlQuery[]; // Some "global" variables needed later on the query (eg. "classified in" subquery)
    queryPart: GeneratedAqlQuery;
}

export interface IRecordRepo {
    createRecord({
        libraryId,
        recordData,
        ctx
    }: {
        libraryId: string;
        recordData: IRecord;
        ctx: IQueryInfos;
    }): Promise<IRecord>;
    updateRecord({
        libraryId,
        recordData,
        mergeObjects = true
    }: {
        libraryId: string;
        recordData: IRecord;
        mergeObjects?: boolean;
    }): Promise<IRecord>;
    deleteRecord({libraryId, recordId, ctx}: {libraryId: string; recordId: string; ctx: IQueryInfos}): Promise<IRecord>;
    find(params: {
        libraryId: string;
        filters?: IRecordFilterOption[];
        sort?: IRecordSort;
        pagination?: IPaginationParams | ICursorPaginationParams;
        withCount?: boolean;
        retrieveInactive?: boolean;
        fulltextSearch?: string;
        ctx: IQueryInfos;
    }): Promise<IListWithCursor<IRecord>>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
    'core.infra.db.dbUtils'?: IDbUtils;
    'core.infra.attributeTypes'?: IAttributeTypesRepo;
    'core.infra.attribute'?: IAttributeRepo;
    'core.infra.attributeTypes.helpers.getConditionPart'?: GetConditionPart;
    'core.infra.record.helpers.getSearchVariablesQueryPart'?: GetSearchVariablesQueryPart;
    'core.infra.record.helpers.getSearchVariableName'?: GetSearchVariableName;
    'core.infra.record.helpers.filterTypes'?: IFilterTypesHelper;
    'core.infra.indexation.helpers.getSearchQuery'?: GetSearchQuery;
}

export default function ({
    'core.infra.db.dbService': dbService = null,
    'core.infra.db.dbUtils': dbUtils = null,
    'core.infra.attributeTypes': attributeTypesRepo = null,
    'core.infra.attributeTypes.helpers.getConditionPart': getConditionPart = null,
    'core.infra.record.helpers.getSearchVariablesQueryPart': getSearchVariablesQueryPart = null,
    'core.infra.record.helpers.getSearchVariableName': getSearchVariableName = null,
    'core.infra.record.helpers.filterTypes': filterTypesHelper = null,
    'core.infra.indexation.helpers.getSearchQuery': getSearchQuery = null,
    'core.infra.attribute': attributeRepo = null
}: IDeps = {}): IRecordRepo {
    const _generateCursor = (from: number, direction: CursorDirection): string =>
        Buffer.from(`${direction}:${from}`).toString('base64');

    const _parseCursor = (
        cursor: string
    ): {
        direction: string;
        from: string;
    } => {
        const s = Buffer.from(cursor, 'base64').toString();
        const [direction, from] = s.split(':');

        return {
            direction,
            from
        };
    };

    return {
        async find({
            libraryId,
            filters,
            sort,
            pagination,
            withCount,
            fulltextSearch,
            retrieveInactive = false,
            ctx
        }): Promise<IListWithCursor<IRecord>> {
            const withCursorPagination = !!pagination && !!(pagination as ICursorPaginationParams).cursor;
            // Force disabling count on cursor  pagination as it's pointless
            const withTotalCount = withCount && !withCursorPagination;
            const coll = dbService.db.collection(libraryId);
            let fulltextSearchQuery: GeneratedAqlQuery;

            if (typeof fulltextSearch !== 'undefined' && fulltextSearch !== '') {
                // format search query
                const cleanFulltextSearch = fulltextSearch?.replace(/\s+/g, ' ').trim();
                const fullTextAttributes = await attributeRepo.getLibraryFullTextAttributes({libraryId, ctx});

                fulltextSearchQuery = getSearchQuery(
                    libraryId,
                    fullTextAttributes.map(a => a.id),
                    cleanFulltextSearch
                );
            }

            const queryParts = [aql`FOR r IN (${fulltextSearchQuery ?? coll})`];

            let isFilteringOnActive = false;

            const aqlPartByOperator: Record<Operator, GeneratedAqlQuery> = {
                [Operator.AND]: aql`AND`,
                [Operator.OR]: aql`OR`,
                [Operator.OPEN_BRACKET]: aql`(`,
                [Operator.CLOSE_BRACKET]: aql`)`
            };

            if (typeof filters !== 'undefined' && filters.length) {
                // Get all variables definitions
                const variablesDeclarations = getSearchVariablesQueryPart(filters);

                const filterStatements: GeneratedAqlQuery[] = [aql`FILTER (`];
                for (const filter of filters) {
                    if (filter.operator) {
                        filterStatements.push(aqlPartByOperator[filter.operator]);
                    } else if (filterTypesHelper.isAttributeFilter(filter)) {
                        isFilteringOnActive = isFilteringOnActive || filter.attributes[0].id === 'active';
                        const variableName = getSearchVariableName(filter);
                        const lastFilterAttribute = filter.attributes.slice(-1)[0];

                        const variableNameAql = literal(variableName);
                        let statement: GeneratedAqlQuery;

                        if (filterTypesHelper.isCountFilter(filter)) {
                            // For count filters, variable only contains the number of values
                            let conditionApplied: AttributeCondition;
                            let valueToCheck = filter.value;
                            switch (filter.condition) {
                                case AttributeCondition.VALUES_COUNT_EQUAL:
                                    conditionApplied = AttributeCondition.EQUAL;
                                    break;
                                case AttributeCondition.VALUES_COUNT_GREATER_THAN:
                                    conditionApplied = AttributeCondition.GREATER_THAN;
                                    break;
                                case AttributeCondition.VALUES_COUNT_LOWER_THAN:
                                    conditionApplied = AttributeCondition.LESS_THAN;
                                    break;
                                case AttributeCondition.IS_EMPTY:
                                    conditionApplied = AttributeCondition.EQUAL;
                                    valueToCheck = 0;
                                    break;
                                case AttributeCondition.IS_NOT_EMPTY:
                                    conditionApplied = AttributeCondition.GREATER_THAN;
                                    valueToCheck = 0;
                            }
                            const countConditionPart = getConditionPart(
                                variableName,
                                conditionApplied,
                                valueToCheck,
                                lastFilterAttribute
                            );
                            statement = aql`${countConditionPart}`;
                        } else {
                            // If multiple values or versionable attribute, apply filter on each value of the array
                            // Otherwise, apply filter on the first value of the array
                            const arrayConditionPart = getConditionPart(
                                'CURRENT',
                                filter.condition as AttributeCondition,
                                filter.value,
                                lastFilterAttribute
                            );

                            const standardConditionPart = getConditionPart(
                                variableName,
                                filter.condition as AttributeCondition,
                                filter.value,
                                lastFilterAttribute
                            );

                            statement = aql`IS_ARRAY(${variableNameAql}) ? LENGTH(${literal(
                                variableNameAql
                            )}[* FILTER ${arrayConditionPart}]) : ${standardConditionPart}`;
                        }

                        filterStatements.push(join([aql`(`, statement, aql`)`]));
                    } else if (filterTypesHelper.isClassifyingFilter(filter)) {
                        const variableName = getSearchVariableName(filter);
                        const classifyingCondition = filter.condition === TreeCondition.CLASSIFIED_IN ? 'IN' : 'NOT IN';
                        filterStatements.push(aql`r._id ${literal(classifyingCondition)} ${literal(variableName)}`);
                    }
                }

                filterStatements.push(aql`)`);

                queryParts.push(join(variablesDeclarations, '\n'));
                queryParts.push(join(filterStatements, '\n'));
            }

            // If we have a full text search query and no specific sort, sorting by relevance is already handled.
            if (sort || !fulltextSearchQuery) {
                const sortQueryPart = sort
                    ? attributeTypesRepo.getTypeRepo(sort.attributes[0]).sortQueryPart(sort)
                    : aql`SORT ${literal('TO_NUMBER(r._key) DESC')}`;

                queryParts.push(sortQueryPart as GeneratedAqlQuery);
            }

            if (!retrieveInactive && !isFilteringOnActive) {
                queryParts.push(aql`FILTER r.active == true`);
            }

            if (pagination) {
                if (!(pagination as IPaginationParams).offset && !(pagination as ICursorPaginationParams).cursor) {
                    (pagination as IPaginationParams).offset = 0;
                }

                if (typeof (pagination as IPaginationParams).offset !== 'undefined') {
                    queryParts.push(aql`LIMIT ${(pagination as IPaginationParams).offset}, ${pagination.limit}`);
                } else if ((pagination as ICursorPaginationParams).cursor) {
                    const {direction, from} = _parseCursor((pagination as ICursorPaginationParams).cursor);

                    // When looking for previous records, first sort in reverse order to get the last records
                    if (direction === CursorDirection.PREV) {
                        queryParts.push(aql`SORT r.created_at ASC, r._key ASC`);
                    }

                    const operator = direction === CursorDirection.NEXT ? '<' : '>';
                    queryParts.push(aql`FILTER r._key ${literal(operator)} ${from}`);
                    queryParts.push(aql`LIMIT ${pagination.limit}`);
                }
            }

            queryParts.push(aql`RETURN MERGE(r, {library: ${libraryId}})`);

            const fullQuery = join(queryParts, '\n');

            const records = await dbService.execute<IExecuteWithCount | IDbDocument[]>({
                query: fullQuery,
                withTotalCount,
                ctx
            });

            const list = withTotalCount ? (records as IExecuteWithCount).results : (records as IDbDocument[]);
            const totalCount = withTotalCount ? (records as IExecuteWithCount).totalCount : null;

            // TODO: detect if we reach end/beginning of the list and should not provide a cursor
            const cursor: IPaginationCursors = pagination
                ? {
                      prev: list.length ? _generateCursor(Number(list[0]._key), CursorDirection.PREV) : null,
                      next: list.length ? _generateCursor(Number(list.slice(-1)[0]._key), CursorDirection.NEXT) : null
                  }
                : null;

            const returnVal = {
                totalCount,
                list: list.map(dbUtils.cleanup),
                cursor
            };

            return returnVal;
        },
        async createRecord({libraryId, recordData, ctx}): Promise<IRecord> {
            const collection = dbService.db.collection(libraryId);
            let newRecord = await collection.save(recordData);
            newRecord = await collection.document(newRecord);

            (newRecord as IRecord).library = newRecord._id.split('/')[0];

            return dbUtils.cleanup(newRecord);
        },
        async deleteRecord({libraryId, recordId, ctx}): Promise<IRecord> {
            const collection = dbService.db.collection(libraryId);

            // Delete record
            const deletedRecord: IRecord = await collection.remove({_key: String(recordId)}, {returnOld: true});

            deletedRecord.library = deletedRecord._id.split('/')[0];
            deletedRecord.old = dbUtils.cleanup(deletedRecord.old);

            return dbUtils.cleanup(deletedRecord);
        },
        async updateRecord({libraryId, recordData, mergeObjects = true}): Promise<IRecord> {
            const collection = dbService.db.collection(libraryId);
            const dataToSave = {...recordData};
            const recordId = dataToSave.id;
            delete dataToSave.id; // Don't save ID

            const updatedRecord: IRecord = await collection.update({_key: String(recordId)}, dataToSave, {
                mergeObjects,
                returnOld: true,
                keepNull: false
            });

            updatedRecord.library = updatedRecord._id.split('/')[0];

            return dbUtils.cleanup(updatedRecord);
        }
    };
}
