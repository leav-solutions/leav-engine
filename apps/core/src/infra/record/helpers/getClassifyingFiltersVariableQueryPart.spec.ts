// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import {IDbService} from 'infra/db/dbService';
import {TreeCondition} from '../../../_types/record';
import {IFilterTypesHelper} from './filterTypes';
import getClassifyingFiltersVariableQueryPart from './getClassifyingFiltersVariableQueryPart';

describe('getClassifyingFiltersVariableQueryPart', () => {
    test('Return variable query part for classifying filters', async () => {
        const mockDbService: Mockify<IDbService> = {
            db: new Database()
        };

        const mockFilterTypesHelper: Mockify<IFilterTypesHelper> = {
            isClassifyingFilter: jest.fn().mockReturnValue(true)
        };

        const func = getClassifyingFiltersVariableQueryPart({
            'core.infra.db.dbService': mockDbService as IDbService,
            'core.infra.record.helpers.filterTypes': mockFilterTypesHelper as IFilterTypesHelper
        });

        const queryPart = func({
            condition: TreeCondition.CLASSIFIED_IN,
            treeId: 'my_tree',
            value: '123456'
        });

        expect(queryPart).toMatchSnapshot();
        expect(queryPart.query).toMatch('OUTBOUND');
    });
});
