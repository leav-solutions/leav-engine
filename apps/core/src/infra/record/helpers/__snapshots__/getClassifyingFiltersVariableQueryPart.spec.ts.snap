// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getClassifyingFiltersVariableQueryPart Return variable query part for classifying filters 1`] = `
{
  "_source": [Function],
  "bindVars": {
    "@value2": "core_edge_tree_my_tree",
    "value0": 1000,
    "value1": "core_nodes_my_tree/123456",
  },
  "query": "
            FOR v, e IN 1..@value0 OUTBOUND @value1
                @@value2
                LET record = DOCUMENT(
                    v.libraryId,
                    v.recordId
                )
                RETURN record._id
        ",
}
`;
